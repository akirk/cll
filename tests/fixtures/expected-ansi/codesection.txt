To ensure [1monly[m the two keys [34m"blogname"[m and [34m"blogdescription"[m are accepted when processing the options, you can add a whitelist check before calling [34mupdate_option[m. Here's how you can modify your code:

[34m[m```php
[90m$allowed_keys = ['blogname', 'blogdescription'];

$blueprint = json_decode('{
	"steps": [
		{
			"step": "setSiteOptions",
			"options": {
				"blogname": "Step Library Demo",
				"blogdescription": "Trying out WordPress Playground."
			}
		}
	]
}', true); // Note: pass true to get associative array instead of stdClass

foreach ($blueprint['steps'] as $step) {
	switch ($step['step']) {
		case "setSiteOptions":
			foreach ($step['options'] as $key => $value) {
				// Only accept allowed keys
				if (in_array($key, $allowed_keys, true)) {
					update_option($key, $value);
				} else {
					// Optionally, handle unexpected keys, e.g., log or throw an error
					error_log("Unexpected option key: {$key}");
				}
			}
			break;
	}
}
[m```

[4mImportant notes:[m
1. [1mUse [34mjson_decode[m with [34mtrue[m as the second parameter[m: This ensures you decode into an associative array, making [34m$blueprint['steps'][m accessible as an array. Otherwise, you'd access objects via [34m->[m.

2. [1mWhitelist keys[m: Only keys that are exactly [34m"blogname"[m or [34m"blogdescription"[m are accepted and passed into [34mupdate_option[m.

3. [1mHandle unexpected keys[m: You might want to log or throw exceptions if invalid keys are present. This helps catch errors or malicious input early.

[4mSummary[m
This approach allows your code to safely accept only those two keys when updating the site options while ignoring or reporting any others.Absolutely! Using [34marray_intersect_key[m is a clean and efficient way to filter your [34m$step['options'][m to only allow certain keys.

Here's how you can do it with your example:

[34m[m```php
[90m$allowed_keys = ['blogname' => true, 'blogdescription' => true];

$blueprint = json_decode('{
	"steps": [
		{
			"step": "setSiteOptions",
			"options": {
				"blogname": "Step Library Demo",
				"blogdescription": "Trying out WordPress Playground."
			}
		}
	]
}', true);

foreach ($blueprint['steps'] as $step) {
	switch ($step['step']) {
		case "setSiteOptions":
			// Filter options to only allowed keys
			$filtered_options = array_intersect_key($step['options'], $allowed_keys);

			foreach ($filtered_options as $key => $value) {
				update_option($key, $value);
			}
			break;
	}
}
[m```

[4mExplanation:[m
- We create [34m$allowed_keys[m as an associative array where keys are what you want to allow.
- [34marray_intersect_key($step['options'], $allowed_keys)[m keeps only the keys in [34m$step['options'][m that also exist in [34m$allowed_keys[m.
- Then you safely update only those allowed options.

This approach is neat and avoids going through and checking each key with [34min_array[m manually.