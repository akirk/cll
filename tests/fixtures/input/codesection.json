[
    "To",
    " ensure",
    " **",
    "only",
    "**",
    " the",
    " two",
    " keys",
    " `\"",
    "blog",
    "name",
    "\"`",
    " and",
    " `\"",
    "blog",
    "description",
    "\"`",
    " are",
    " accepted",
    " when",
    " processing",
    " the",
    " options",
    ",",
    " you",
    " can",
    " add",
    " a",
    " whitelist",
    " check",
    " before",
    " calling",
    " `",
    "update",
    "_option",
    "`.",
    " Here's",
    " how",
    " you",
    " can",
    " modify",
    " your",
    " code",
    ":\n\n",
    "```",
    "php",
    "\n",
    "$",
    "allowed",
    "_keys",
    " =",
    " ['",
    "blog",
    "name",
    "',",
    " '",
    "blog",
    "description",
    "'];\n\n",
    "$",
    "blue",
    "print",
    " =",
    " json",
    "_decode",
    "('",
    "{\n",
    " ",
    " \"",
    "steps",
    "\":",
    " [\n",
    "   ",
    " {\n",
    "     ",
    " \"",
    "step",
    "\":",
    " \"",
    "set",
    "Site",
    "Options",
    "\",\n",
    "     ",
    " \"",
    "options",
    "\":",
    " {\n",
    "       ",
    " \"",
    "blog",
    "name",
    "\":",
    " \"",
    "Step",
    " Library",
    " Demo",
    "\",\n",
    "       ",
    " \"",
    "blog",
    "description",
    "\":",
    " \"",
    "Trying",
    " out",
    " Word",
    "Press",
    " Playground",
    ".\"\n",
    "     ",
    " }\n",
    "   ",
    " }\n",
    " ",
    " ]\n",
    "}',",
    " true",
    ");",
    " //",
    " Note",
    ":",
    " pass",
    " true",
    " to",
    " get",
    " associative",
    " array",
    " instead",
    " of",
    " std",
    "Class",
    "\n\n",
    "foreach",
    " ($",
    "blue",
    "print",
    "['",
    "steps",
    "']",
    " as",
    " $",
    "step",
    ")",
    " {\n",
    " ",
    " switch",
    " ($",
    "step",
    "['",
    "step",
    "'])",
    " {\n",
    "   ",
    " case",
    " \"",
    "set",
    "Site",
    "Options",
    "\":\n",
    "     ",
    " foreach",
    " ($",
    "step",
    "['",
    "options",
    "']",
    " as",
    " $",
    "key",
    " =>",
    " $",
    "value",
    ")",
    " {\n",
    "       ",
    " //",
    " Only",
    " accept",
    " allowed",
    " keys",
    "\n",
    "       ",
    " if",
    " (",
    "in",
    "_array",
    "($",
    "key",
    ",",
    " $",
    "allowed",
    "_keys",
    ",",
    " true",
    "))",
    " {\n",
    "         ",
    " update",
    "_option",
    "($",
    "key",
    ",",
    " $",
    "value",
    ");\n",
    "       ",
    " }",
    " else",
    " {\n",
    "         ",
    " //",
    " Option",
    "ally",
    ",",
    " handle",
    " unexpected",
    " keys",
    ",",
    " e",
    ".g",
    ".,",
    " log",
    " or",
    " throw",
    " an",
    " error",
    "\n",
    "         ",
    " error",
    "_log",
    "(\"",
    "Unexpected",
    " option",
    " key",
    ":",
    " {$",
    "key",
    "}\");\n",
    "       ",
    " }\n",
    "     ",
    " }\n",
    "     ",
    " break",
    ";\n",
    " ",
    " }\n",
    "}\n",
    "``",
    "`\n\n",
    "###",
    " Important",
    " notes",
    ":\n",
    "1",
    ".",
    " **",
    "Use",
    " `",
    "json",
    "_decode",
    "`",
    " with",
    " `",
    "true",
    "`",
    " as",
    " the",
    " second",
    " parameter",
    "**",
    ":",
    " This",
    " ensures",
    " you",
    " decode",
    " into",
    " an",
    " associative",
    " array",
    ",",
    " making",
    " `$",
    "blue",
    "print",
    "['",
    "steps",
    "']",
    "`",
    " accessible",
    " as",
    " an",
    " array",
    ".",
    " Otherwise",
    ",",
    " you'd",
    " access",
    " objects",
    " via",
    " `",
    "->",
    "`.\n\n",
    "2",
    ".",
    " **",
    "Whitelist",
    " keys",
    "**",
    ":",
    " Only",
    " keys",
    " that",
    " are",
    " exactly",
    " `\"",
    "blog",
    "name",
    "\"`",
    " or",
    " `\"",
    "blog",
    "description",
    "\"`",
    " are",
    " accepted",
    " and",
    " passed",
    " into",
    " `",
    "update",
    "_option",
    "`.\n\n",
    "3",
    ".",
    " **",
    "Handle",
    " unexpected",
    " keys",
    "**",
    ":",
    " You",
    " might",
    " want",
    " to",
    " log",
    " or",
    " throw",
    " exceptions",
    " if",
    " invalid",
    " keys",
    " are",
    " present",
    ".",
    " This",
    " helps",
    " catch",
    " errors",
    " or",
    " malicious",
    " input",
    " early",
    ".\n\n",
    "###",
    " Summary",
    "\n",
    "This",
    " approach",
    " allows",
    " your",
    " code",
    " to",
    " safely",
    " accept",
    " only",
    " those",
    " two",
    " keys",
    " when",
    " updating",
    " the",
    " site",
    " options",
    " while",
    " ignoring",
    " or",
    " reporting",
    " any",
    " others",
    ".",
    "Absolutely",
    "!",
    " Using",
    " `",
    "array",
    "_inter",
    "sect",
    "_key",
    "`",
    " is",
    " a",
    " clean",
    " and",
    " efficient",
    " way",
    " to",
    " filter",
    " your",
    " `$",
    "step",
    "['",
    "options",
    "']",
    "`",
    " to",
    " only",
    " allow",
    " certain",
    " keys",
    ".\n\n",
    "Here's",
    " how",
    " you",
    " can",
    " do",
    " it",
    " with",
    " your",
    " example",
    ":\n\n",
    "```",
    "php",
    "\n",
    "$",
    "allowed",
    "_keys",
    " =",
    " ['",
    "blog",
    "name",
    "'",
    " =>",
    " true",
    ",",
    " '",
    "blog",
    "description",
    "'",
    " =>",
    " true",
    "];\n\n",
    "$",
    "blue",
    "print",
    " =",
    " json",
    "_decode",
    "('",
    "{\n",
    " ",
    " \"",
    "steps",
    "\":",
    " [\n",
    "   ",
    " {\n",
    "     ",
    " \"",
    "step",
    "\":",
    " \"",
    "set",
    "Site",
    "Options",
    "\",\n",
    "     ",
    " \"",
    "options",
    "\":",
    " {\n",
    "       ",
    " \"",
    "blog",
    "name",
    "\":",
    " \"",
    "Step",
    " Library",
    " Demo",
    "\",\n",
    "       ",
    " \"",
    "blog",
    "description",
    "\":",
    " \"",
    "Trying",
    " out",
    " Word",
    "Press",
    " Playground",
    ".\"\n",
    "     ",
    " }\n",
    "   ",
    " }\n",
    " ",
    " ]\n",
    "}',",
    " true",
    ");\n\n",
    "foreach",
    " ($",
    "blue",
    "print",
    "['",
    "steps",
    "']",
    " as",
    " $",
    "step",
    ")",
    " {\n",
    " ",
    " switch",
    " ($",
    "step",
    "['",
    "step",
    "'])",
    " {\n",
    "   ",
    " case",
    " \"",
    "set",
    "Site",
    "Options",
    "\":\n",
    "     ",
    " //",
    " Filter",
    " options",
    " to",
    " only",
    " allowed",
    " keys",
    "\n",
    "     ",
    " $",
    "filtered",
    "_options",
    " =",
    " array",
    "_inter",
    "sect",
    "_key",
    "($",
    "step",
    "['",
    "options",
    "'],",
    " $",
    "allowed",
    "_keys",
    ");\n\n",
    "     ",
    " foreach",
    " ($",
    "filtered",
    "_options",
    " as",
    " $",
    "key",
    " =>",
    " $",
    "value",
    ")",
    " {\n",
    "       ",
    " update",
    "_option",
    "($",
    "key",
    ",",
    " $",
    "value",
    ");\n",
    "     ",
    " }\n",
    "     ",
    " break",
    ";\n",
    " ",
    " }\n",
    "}\n",
    "``",
    "`\n\n",
    "###",
    " Explanation",
    ":\n",
    "-",
    " We",
    " create",
    " `$",
    "allowed",
    "_keys",
    "`",
    " as",
    " an",
    " associative",
    " array",
    " where",
    " keys",
    " are",
    " what",
    " you",
    " want",
    " to",
    " allow",
    ".\n",
    "-",
    " `",
    "array",
    "_inter",
    "sect",
    "_key",
    "($",
    "step",
    "['",
    "options",
    "'],",
    " $",
    "allowed",
    "_keys",
    ")`",
    " keeps",
    " only",
    " the",
    " keys",
    " in",
    " `$",
    "step",
    "['",
    "options",
    "']",
    "`",
    " that",
    " also",
    " exist",
    " in",
    " `$",
    "allowed",
    "_keys",
    "`.\n",
    "-",
    " Then",
    " you",
    " safely",
    " update",
    " only",
    " those",
    " allowed",
    " options",
    ".\n\n",
    "This",
    " approach",
    " is",
    " neat",
    " and",
    " avoids",
    " going",
    " through",
    " and",
    " checking",
    " each",
    " key",
    " with",
    " `",
    "in",
    "_array",
    "`",
    " manually",
    "."
]