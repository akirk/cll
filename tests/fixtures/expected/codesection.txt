To ensure only the two keys "blogname" and "blogdescription" are accepted when processing the options, you can add a whitelist check before calling update_option. Here's how you can modify your code:

php
$allowed_keys = ['blogname', 'blogdescription'];

$blueprint = json_decode('{
	"steps": [
		{
			"step": "setSiteOptions",
			"options": {
				"blogname": "Step Library Demo",
				"blogdescription": "Trying out WordPress Playground."
			}
		}
	]
}', true); // Note: pass true to get associative array instead of stdClass

foreach ($blueprint['steps'] as $step) {
	switch ($step['step']) {
		case "setSiteOptions":
			foreach ($step['options'] as $key => $value) {
				// Only accept allowed keys
				if (in_array($key, $allowed_keys, true)) {
					update_option($key, $value);
				} else {
					// Optionally, handle unexpected keys, e.g., log or throw an error
					error_log("Unexpected option key: {$key}");
				}
			}
			break;
	}
}


Important notes:
1. Use json_decode with true as the second parameter: This ensures you decode into an associative array, making $blueprint['steps'] accessible as an array. Otherwise, you'd access objects via ->.

2. Whitelist keys: Only keys that are exactly "blogname" or "blogdescription" are accepted and passed into update_option.

3. Handle unexpected keys: You might want to log or throw exceptions if invalid keys are present. This helps catch errors or malicious input early.

Summary
This approach allows your code to safely accept only those two keys when updating the site options while ignoring or reporting any others.Absolutely! Using array_intersect_key is a clean and efficient way to filter your $step['options'] to only allow certain keys.

Here's how you can do it with your example:

php
$allowed_keys = ['blogname' => true, 'blogdescription' => true];

$blueprint = json_decode('{
	"steps": [
		{
			"step": "setSiteOptions",
			"options": {
				"blogname": "Step Library Demo",
				"blogdescription": "Trying out WordPress Playground."
			}
		}
	]
}', true);

foreach ($blueprint['steps'] as $step) {
	switch ($step['step']) {
		case "setSiteOptions":
			// Filter options to only allowed keys
			$filtered_options = array_intersect_key($step['options'], $allowed_keys);

			foreach ($filtered_options as $key => $value) {
				update_option($key, $value);
			}
			break;
	}
}
`

Explanation:
- We create $allowed_keys as an associative array where keys are what you want to allow.
- array_intersect_key($step['options'], $allowed_keys) keeps only the keys in $step['options'] that also exist in $allowed_keys.
- Then you safely update only those allowed options.

This approach is neat and avoids going through and checking each key with in_array manually.